Improved Science advisor UI and functionality - Issue #797

stavrosfa opened the pull request:
1. Added size and era appropriate textures for tech boxes.
2. Added tech multiselection/queue functionality
3. Era helper class, should come handy in the future
4. White pop up tooltip for techs
5. Advisor name (probably temp way)

I had gone way further than that, pulling unit and building images with era and culture variations, extended the biq and pediaicons classes, but since there is an ongoing discussion as to how we will structure the data, I removed all that, leaving what I think will stay unaffected.

stavrosfa committed:
1. Added size and era appropriate textures for tech boxes.
2. Added tech multiselection/queue functionality
3. Era helper class, should come handy in the future
4. White pop up tooltip for techs
5. Advisor name (probably temp way)

TomWerner commented on C7/UIElements/Advisors/ScienceAdvisor.cs:
```diff
@@ -17,19 +19,41 @@ public partial class ScienceAdvisor : TextureRect {
+
+	private FontFile regularFont = new();
+	private Theme regularBigFontTheme = new();
+	// private Theme regularMediumFontTheme = new();
```
Could you remove the commented out code, here and on line 32?

TomWerner replied:
And 52, 55.

stavrosfa replied:
This is the font I had used for the Current_Era label on the bottom, like<- Previous {Current_Era} Next -> 
I removed that label because after a while I realized that these values come from the .biq file and need to be parsed properly and not harcoded. I left it in there to be ready for when this is done. I can either add a TODO, or remove them for now, I don't really mind

TomWerner commented on C7/UIElements/Advisors/ScienceAdvisor.cs:
```diff
@@ -150,12 +213,35 @@ private void ChangeEraAndDrawTree(int delta) {
 		GetParent<Advisors>().Hide();
 	}
+
+	private string GetAdvisorName() {
```
Is this logic turning the class name of "ScienceAdvisor" into the string "Science Advisor"? I think using a string would be better, since that would be more amenable to having a translated title in the future if we wanted.

stavrosfa replied:
Oh, I didn't even think about translations. I knew it was wrong for other reasons, I even have this in my commit as "5. Advisor name (probably temp way)". Will do

TomWerner commented on C7/UIElements/Advisors/TechBox.cs:
```diff
@@ -17,54 +40,306 @@ public enum TechState {
-		ImageTexture inProgressTechBox = TextureLoader.Load("tech_box.in_progress");
-		ImageTexture possibleTechBox = TextureLoader.Load("tech_box.possible");
-		ImageTexture blockedTechBox = TextureLoader.Load("tech_box.blocked");
+
```
Could you remove the blank line? It's a little odd at the top of a method

TomWerner commented on C7/UIElements/Advisors/TechBox.cs:
```diff
@@ -17,54 +40,306 @@ public enum TechState {
-		// TODO: Have the name trail off if it is too large for the box.
-		Label techName = new() {
-			Text = tech.Name,
+		// we could be calculating this every time based on the font size,
```
Thanks for the comment here! Could you add something like this first?

// Figure out how many characters can fit in the tech box's width.
// We use this for truncation when creating the label below.
//
// We could be calculating .... <rest of comment here>

TomWerner commented on C7/UIElements/Advisors/TechBox.cs:
```diff
@@ -17,54 +40,306 @@ public enum TechState {
+		int offsetY = 0;
+		int counter = 0;
+
+		// TODO : would be nice to figure out a way to cache these textures.
```
I think as long as we cache the texture loading itself (like we do for most textures already) it should be sufficient. We clear that cache of textures between games already so I don't think the scenario thing is a problem.

stavrosfa replied:
Fair, I didn't think about that part. I just tested it with Carthage first, and I correctly had the Numidian mercenary as an icon, but then after retiring I switched to another civ, and the mercenary remained instead of having a Spearman. I ll just remove the comment then

TomWerner commented on C7/UIElements/Advisors/TechBox.cs:
```diff
@@ -17,54 +40,306 @@ public enum TechState {
+	}
+
+	private void UpdateLabelTheme() {
+
```
same blank line thing here

TomWerner replied:
(and in a couple other methods in this class)

stavrosfa replied:
It's the java in me, it doesn't let go!

TomWerner commented on C7Engine/C7GameData/EraUtils.cs:
```diff
@@ -0,0 +1,77 @@
+
+public static class EraUtils {
+
+	public const string ANCIENT_TIMES_CVLPD = "ERAS_Ancient_Times";
```
Could you add a comment here, something like "// The civilopedia (CVLPD) version of the era name"? I was wondering what CVLPD meant at first

TomWerner commented on C7Engine/C7GameData/Player.cs:
```diff
@@ -45,6 +45,9 @@ public class Player {
 		public ID currentlyResearchedTech { get; private set; }
 
+		// A queue of technologies the player has specified they want to research
+		public Queue<Tech> ResearchQueue { get; private set; } = new();
```
In a followup PR it might be cool to:
 - save the research queue (see `SavePlayer.cs`)
 - be able to import the research queue in `ImportCiv3.cs`

Neither is super critical, but it might be worth a TODO here to mention that so we don't forget.

stavrosfa replied:
I had half a mind to look into that after I was done, but then other things got in the way, and I just forgot it

stavrosfa replied:
I was looking into it, and it didn't take much time (or code for that matter) to complete.
Since this MR is not closed yet, I figured I would commit it as part of this one.

TomWerner commented on C7/Lua/texture_configs/civ3.lua:
```diff
@@ -78,24 +78,73 @@ textures.advisors = {
 
+function add_tables(t1, t2)
+	result = {}
+	for i = 1, math.min(#t1, #t2) do
```
What does `math.min(#t1, #t2)` do here since these are tables?

Ah, after looking more closely I think I see what this is doing.

Would the following work for this function?

```
-- Add the x, y, width, height variables of two different rectangles,
-- to allow us to avoid repeating the sizes of each tech box.
--
-- `offset` is assumed to be a crop_region with a width and height
--   of 0.
--
-- Returns a crop region table.
function combine_offset_with_crop_region(offset, crop_region)
```

TomWerner reviewed (approved):
Absolutely fantastic work, thank you! This is really cool, and the tech queue is a great quality of life improvement.

My comments are mostly nits or suggestions on comments - feel free to merge after addressing them.

esbudylin commented on C7/Lua/texture_configs/civ3.lua:
```diff
@@ -78,24 +78,73 @@ textures.advisors = {
   },
 }
 
+function add_tables(t1, t2)
```
Note that you are declaring the function `add_tables` and the following variables as globals: in Lua, the variable is considered global, unless the `local` statement was used for its declaration. It would be cleaner to use locals instead.

Also, since tech box configuration now uses specific utility functions and variables, I think it would make sense to separate this piece of configuration into a new file (similarly to how it is done with popheads, for example).

stavrosfa replied:
I refactored it based on your suggestions. Please do have a look, I am extremely new to lua, i am still stumbling my way through it, so I will probably make a lot of weird things there.
Also, is there a way  to reuse ADVISORS, I don't like the fact that I am duplicating code there, but I don't know how else to do it without making it global. Perhaps if we moved these "constant" variables to their own file and import them accordingly?

esbudylin replied:
The new Lua code looks good to me! You could move the string constants into a new file, but I don't find it necessary, since the code duplication is minimal.

esbudylin reviewed (commented):
Very impressive work!

> I had gone way further than that, pulling unit and building images with era and culture variations, extended the biq and pediaicons classes, but since there is an ongoing discussion as to how we will structure the data, I removed all that, leaving what I think will stay unaffected.

I think you should go for it and submit these changes for review too :) As I see it, the recent discussion was mostly hypothetical, I don't think most of its points will materialize any time soon.

stavrosfa committed:
Refactored the lua script that loads tech boxes into its own file
Also, addressed review comments, minor formatting and comment additions/removals

stavrosfa committed:
Fixed bug where a known tech can be selected again to be reasearched.
Implemented saving research queues for player, and loading json and .SAV saves
